File: src/js/core/repositories.js
/* Typed frontend repository facade.
 *
 * Feature modules call Graphite.repositories.<domain> rather than invoking
 * Tauri commands directly. Browser builds remain untouched: callers can use
 * isAvailable() to keep IndexedDB as the compatibility path during migration.
 */
(function () {
  const Graphite = window.Graphite = window.Graphite || {};

  function native(command, args) {
    if (!Graphite.native?.available) throw new Error('SQLite repositories require the Tauri runtime.');
    return Graphite.native.invoke(command, args);
  }

  const softDelete = (command) => (id, hard = false) => native(command, { id, hard });

  Graphite.repositories = {
    isAvailable: () => Boolean(Graphite.native?.available && Graphite.database?.isAvailable?.()),
    workspaces: {
      list: (includeArchived = false) => native('repository_workspace_list', { includeArchived }),
      get: (id) => native('repository_workspace_get', { id }),
      upsert: (input) => native('repository_workspace_upsert', { input }),
      remove: (id) => native('repository_workspace_delete', { id })
    },
    notes: {
      list: (workspaceId, includeDeleted = false) => native('repository_note_list', { workspaceId, includeDeleted }),
      get: (id) => native('repository_note_get', { id }),
      upsert: (input) => native('repository_note_upsert', { input }),
      remove: softDelete('repository_note_delete')
    },
    tasks: {
      list: (workspaceId, includeDeleted = false) => native('repository_task_list', { workspaceId, includeDeleted }),
      get: (id) => native('repository_task_get', { id }),
      upsert: (input) => native('repository_task_upsert', { input }),
      remove: softDelete('repository_task_delete')
    },
    calendar: {
      list: (workspaceId, fromAt = null, toAt = null, includeDeleted = false) => native('repository_calendar_list', { workspaceId, fromAt, toAt, includeDeleted }),
      get: (id) => native('repository_calendar_get', { id }),
      upsert: (input) => native('repository_calendar_upsert', { input }),
      remove: softDelete('repository_calendar_delete')
    },
    flashcards: {
      list: (workspaceId, dueBefore = null, includeDeleted = false) => native('repository_flashcard_list', { workspaceId, dueBefore, includeDeleted }),
      get: (id) => native('repository_flashcard_get', { id }),
      upsert: (input) => native('repository_flashcard_upsert', { input }),
      remove: softDelete('repository_flashcard_delete'),
      recordReview: (input) => native('repository_flashcard_review', { input })
    },
    sessions: {
      list: (workspaceId, fromAt = null, toAt = null) => native('repository_session_list', { workspaceId, fromAt, toAt }),
      get: (id) => native('repository_session_get', { id }),
      upsert: (input) => native('repository_session_upsert', { input })
    }
  };
})();

File: src-tauri/src/commands/repositories.rs
use super::super::repositories;
use tauri::AppHandle;

#[tauri::command]
pub fn repository_workspace_list(app: AppHandle, include_archived: Option<bool>) -> Result<Vec<repositories::Workspace>, String> { repositories::workspace_list(&app, include_archived.unwrap_or(false)) }
#[tauri::command]
pub fn repository_workspace_get(app: AppHandle, id: String) -> Result<Option<repositories::Workspace>, String> { repositories::workspace_get(&app, &id) }
#[tauri::command]
pub fn repository_workspace_upsert(app: AppHandle, input: repositories::WorkspaceInput) -> Result<repositories::Workspace, String> { repositories::workspace_upsert(&app, input) }
#[tauri::command]
pub fn repository_workspace_delete(app: AppHandle, id: String) -> Result<bool, String> { repositories::workspace_delete(&app, &id) }

#[tauri::command]
pub fn repository_note_list(app: AppHandle, workspace_id: String, include_deleted: Option<bool>) -> Result<Vec<repositories::Note>, String> { repositories::note_list(&app, &workspace_id, include_deleted.unwrap_or(false)) }
#[tauri::command]
pub fn repository_note_get(app: AppHandle, id: String) -> Result<Option<repositories::Note>, String> { repositories::note_get(&app, &id) }
#[tauri::command]
pub fn repository_note_upsert(app: AppHandle, input: repositories::NoteInput) -> Result<repositories::Note, String> { repositories::note_upsert(&app, input) }
#[tauri::command]
pub fn repository_note_delete(app: AppHandle, id: String, hard: Option<bool>) -> Result<bool, String> { repositories::note_delete(&app, &id, hard.unwrap_or(false)) }

#[tauri::command]
pub fn repository_task_list(app: AppHandle, workspace_id: String, include_deleted: Option<bool>) -> Result<Vec<repositories::Task>, String> { repositories::task_list(&app, &workspace_id, include_deleted.unwrap_or(false)) }
#[tauri::command]
pub fn repository_task_get(app: AppHandle, id: String) -> Result<Option<repositories::Task>, String> { repositories::task_get(&app, &id) }
#[tauri::command]
pub fn repository_task_upsert(app: AppHandle, input: repositories::TaskInput) -> Result<repositories::Task, String> { repositories::task_upsert(&app, input) }
#[tauri::command]
pub fn repository_task_delete(app: AppHandle, id: String, hard: Option<bool>) -> Result<bool, String> { repositories::task_delete(&app, &id, hard.unwrap_or(false)) }

#[tauri::command]
pub fn repository_calendar_list(app: AppHandle, workspace_id: String, from_at: Option<i64>, to_at: Option<i64>, include_deleted: Option<bool>) -> Result<Vec<repositories::CalendarEvent>, String> { repositories::event_list(&app, &workspace_id, from_at, to_at, include_deleted.unwrap_or(false)) }
#[tauri::command]
pub fn repository_calendar_get(app: AppHandle, id: String) -> Result<Option<repositories::CalendarEvent>, String> { repositories::event_get(&app, &id) }
#[tauri::command]
pub fn repository_calendar_upsert(app: AppHandle, input: repositories::CalendarEventInput) -> Result<repositories::CalendarEvent, String> { repositories::event_upsert(&app, input) }
#[tauri::command]
pub fn repository_calendar_delete(app: AppHandle, id: String, hard: Option<bool>) -> Result<bool, String> { repositories::event_delete(&app, &id, hard.unwrap_or(false)) }

#[tauri::command]
pub fn repository_flashcard_list(app: AppHandle, workspace_id: String, due_before: Option<i64>, include_deleted: Option<bool>) -> Result<Vec<repositories::Flashcard>, String> { repositories::card_list(&app, &workspace_id, due_before, include_deleted.unwrap_or(false)) }
#[tauri::command]
pub fn repository_flashcard_get(app: AppHandle, id: String) -> Result<Option<repositories::Flashcard>, String> { repositories::card_get(&app, &id) }
#[tauri::command]
pub fn repository_flashcard_upsert(app: AppHandle, input: repositories::FlashcardInput) -> Result<repositories::Flashcard, String> { repositories::card_upsert(&app, input) }
#[tauri::command]
pub fn repository_flashcard_delete(app: AppHandle, id: String, hard: Option<bool>) -> Result<bool, String> { repositories::card_delete(&app, &id, hard.unwrap_or(false)) }
#[tauri::command]
pub fn repository_flashcard_review(app: AppHandle, input: repositories::FlashcardReviewInput) -> Result<repositories::FlashcardReview, String> { repositories::card_review(&app, input) }

#[tauri::command]
pub fn repository_session_list(app: AppHandle, workspace_id: String, from_at: Option<i64>, to_at: Option<i64>) -> Result<Vec<repositories::FocusSession>, String> { repositories::session_list(&app, &workspace_id, from_at, to_at) }
#[tauri::command]
pub fn repository_session_get(app: AppHandle, id: String) -> Result<Option<repositories::FocusSession>, String> { repositories::session_get(&app, &id) }
#[tauri::command]
pub fn repository_session_upsert(app: AppHandle, input: repositories::FocusSessionInput) -> Result<repositories::FocusSession, String> { repositories::session_upsert(&app, input) }

#[tauri::command]
pub fn repository_remaining_upsert(app: AppHandle, state: repositories::remaining::RemainingState) -> Result<(), String> {
    let workspace_ids = repositories::workspace_list(&app, true)?.into_iter().map(|w| w.id).collect::<std::collections::BTreeSet<_>>();
    repositories::remaining::upsert_state(&app, &workspace_ids, &state)
}

#[tauri::command]
pub fn repository_remaining_snapshot(app: AppHandle, workspace_id: Option<String>) -> Result<repositories::remaining::RemainingState, String> {
    repositories::remaining::snapshot(&app, workspace_id.as_deref())
}


#[tauri::command]
pub fn repository_remaining_initialized(app: AppHandle) -> Result<bool, String> {
    repositories::remaining::initialized(&app)
}
